### DISCUSSION TITLE
Generic State Machines?

### DISCUSSION THREAD

**Discussion by geekox86** [statelyai/xstate#2359] [Q&A] [answered]
Greetings all.

What are your thoughts on generic state machines?
For example:
ToggleMachine (on / off)
VisibilityMachibe (shown / hidden)
GateMachine (open / closed)

All such machines have 2 opposite states.

Would it be wise to have a machine factory where you pass the name of the machine plus its specific 2 states to create a specific machine definition?

I wish there was a parameteric/generic machine for such cases.

What are your thoughts?

**Comment by davidkpiano** [comment, marked as answer]
> Would it be wise to have a machine factory where you pass the name of the machine plus its specific 2 states to create a specific machine definition?

Sure, you can do that.

> I wish there was a parameteric/generic machine for such cases.

What would this look like, for you?

**Reply by geekox86** [reply]
I thought about it for sometime and I could only think of two approaches:

1- some string mapping passed to the options of createMachine

2- some TypeScript type mapping passed as a generic parameter to createMachine but not sure how

**Reply by johnlombardo-dev** [reply]
I feel like by the time you've passed in the machine name, state names, event names, initial state, and potentially some guard conditions you could just as easily have structured those variables in a JSON object... and then you have a machine definition!

To speed up development I would be more inclined to lean on autocomplete code snippets for common patterns. And maybe one day Copilot for state machines 😆

**Comment by jashworth** [comment]
related: https://gist.github.com/jashworth/24f98f0f6b042c194f8de197e9cc6056

I was aiming for a machine that was invariant wrt to the type of the item in the context, and wanted a way to be able to create a typed version of the machine for a type of item determined only at runtime.

Ran into some "interesting" type issues on the way (partly due to React) and could only find a 😱 way around them.

Is there a better approach?
